### PR TITLE
Suggest not requesting one-by-one in rule 2.1

### PR DESCRIPTION
**Pull request by ktoso**
Follow up from https://github.com/reactive-streams/reactive-streams-jvm/pull/433#issuecomment-398727949

Improves intent wording to explain that implementations should try to request as many elements as it makes sense, and not only request(1) as it defeats the pipelining potential of reactive streams.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -131,7 +131,7 @@ public interface Subscriber<T> {
 | ID                        | Rule                                                                                                   |
 | ------------------------- | ------------------------------------------------------------------------------------------------------ |
 | <a name="2.1">1</a>       | A `Subscriber` MUST signal demand via `Subscription.request(long n)` to receive `onNext` signals. |
-| [:bulb:](#2.1 "2.1 explained") | *The intent of this rule is to establish that it is the responsibility of the Subscriber to signal when, and how many, elements it is able and willing to receive.* |
+| [:bulb:](#2.1 "2.1 explained") | *The intent of this rule is to establish that it is the responsibility of the Subscriber to signal when, and how many, elements it is able and willing to receive. It is RECOMMENDED to request as many elements as a Subscriber is prepared to receive (e.g. has preallocated buffer space for), rather than requesting "one by one".* |
```
[the remark this answers is not in the document]
Comment (ktoso) [reply README.md]: Good point, I don't like starting with "for performance reasons", always feels hand wavy right away, but good rewording; how about: 

> It is RECOMMENDED to request as many elements as a Subscriber is currently able to receive, as the most conservative strategy to signal demand by requesting elements one-at-a-time is the inherently inefficient.

?

**Comment by DougLea** [comment]
Or maybe:

It is RECOMMENDED that Subscribers request as many elements as can be processed without loss, within their resource limits. Requesting only one item at a time results in an inherently inefficient "stop-and-wait" protocol.

Maybe we could add a link to something comparing stop-and-wait to windowing. All the ones that come to mind assume people know about TCP protocols etc, which might not be helpful.

**Comment by ktoso** [comment]
I like that wording a lot @DougLea, thanks! 
I think it is good even without linking to more stop-and-wait definitions (e.g. the https://en.wikipedia.org/wiki/Stop-and-wait_ARQ derails quickly into cases with "ack being lost" and other scenarios that could/would confuse people in the context of RS).

Would it be fine if I use your proposed wording in this PR?

**Comment by viktorklang** [comment]
@DougLea @ktoso Great feedback!

Could it perhaps be distilled further, something along the lines of:

>It is RECOMMENDED that Subscribers request the upper limit of what they are able to process as requesting only one item at a time results in an inherently inefficient "stop-and-wait" protocol.

**Comment by ktoso** [comment]
Sounds find to me. The "without loss" phrasing is not strictly necessary I think (I mean it could be a dropping / filtering operation after all)

**Comment by viktorklang** [comment]
@ktoso Right, and some operators do not want too trigger much look-ahead (a headOption / findFirst).

**Comment by viktorklang** [comment]
@ktoso Rebase and resolve conflict and I'll merge. :)

**Comment by ktoso** [comment]
resolved

**Comment by viktorklang** [comment]
Perfect, thanks @ktoso!

**Pull request merged**